subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/api/ArtifactCollectionResultProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,147 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.api
+
+import org.gradle.integtests.fixtures.AbstractHttpDependencyResolutionTest
+import org.gradle.integtests.fixtures.extensions.FluidDependenciesResolveInterceptor
+import org.gradle.integtests.fixtures.extensions.FluidDependenciesResolveTest
+
+@FluidDependenciesResolveTest
+class ArtifactCollectionResultProviderIntegrationTest extends AbstractHttpDependencyResolutionTest {
+
+    def setup() {
+        settingsFile << """
+            rootProject.name = 'root'
+            include 'project-lib'
+        """
+        mavenRepo.module("org.external", "external-lib").publish()
+        file('lib/file-lib.jar') << 'content'
+        buildFile << """
+            project(':project-lib') {
+                apply plugin: 'java'
+            }
+            configurations {
+                compile
+            }
+            repositories {
+                maven { url "${mavenRepo.uri}" }
+            }
+            dependencies {
+                compile 'org.external:external-lib:1.0'
+                compile project('project-lib')
+                compile files('lib/file-lib.jar')
+            }
+
+            abstract class TaskWithArtifactCollectionResultProviderInput extends DefaultTask {
+
+                @InputFiles
+                abstract ConfigurableFileCollection getArtifactFiles()
+
+                @Internal
+                abstract SetProperty<ResolvedArtifactResult> getResolvedArtifacts()
+
+                @OutputFile
+                abstract RegularFileProperty getOutputFile()
+
+                @TaskAction
+                void action() {
+                    println(artifactFiles.files)
+                    println(resolvedArtifacts.get())
+                }
+            }
+        """
+    }
+
+    def "result provider has artifact files and metadata"() {
+        given:
+        buildFile << """
+            tasks.register('checkArtifacts', TaskWithArtifactCollectionResultProviderInput) {
+                artifactFiles.from(configurations.compile.incoming.artifacts.artifactFiles)
+                resolvedArtifacts.set(configurations.compile.incoming.artifacts.resolvedArtifacts)
+                outputFile.set(layout.buildDirectory.file('output.txt'))
+                doLast {
+                    def artifactResults = resolvedArtifacts.get()
+
+                    assert artifactResults.size() == 3
+
+                    // Check external artifact
+                    def idx = artifactFiles.findIndexOf { it.name == 'external-lib-1.0.jar' }
+
+                    def result = artifactResults[idx]
+                    assert result.file == artifactFiles[idx]
+
+                    assert result.id instanceof org.gradle.internal.component.external.model.ModuleComponentArtifactIdentifier
```

Comment:
Should verify that the variant information is also available in the result.

Reply:
Done